Remove batch class get_mask_invalid_actions_forward() calls

carriepl opened the pull request:
This PR aims to reduce the number of calls to the `get_mask_invalid_actions_forward()` method of environment classes. 

Currently, whenever the GFN takes an action on an environment, it involves computing a mask of invalid actions, sampling one of the non-invalid actions, and calling `Batch.add_to_batch()` which will compute the mask of invalid actions a second time. With this PR, the mask of invalid actions is computed only once and provided as an optional argument to `add_to_batch()`.

I tested on the Tetris environments (which, at larger env sizes, has its runtime completely dominated by calls to `get_mask_invalid_actions_forward()` ) with the following command : 
`python main.py user=$USER env=tetris proxy=tetris gflownet=trajectorybalance env.width=10 env.height=10 logger.do.online=False`

Without this PR : ~6.0 seconds/iteration
With this PR : ~4.1 seconds/iteration

alexhernandezgarcia reviewed (approved):
Awesome, thanks! It all looks good to me. [PR #111](https://github.com/alexhernandezgarcia/gflownet/pull/111) is already merged and there were a few changes after the creation of your branch, so there will be a couple of conflicts related to changes of variable names and typing in arguments.

carriepl commented:
With the Batch class merged, I rebased on top of the current main and redid my timings (the Batch branch I used previously didn't have a lot of the speedups that were recently merged into main).
@alexhernandezgarcia Since I had to change the code to solve conflicts and also added some comments, can you spare a look even though you already approved it?

alexhernandezgarcia reviewed (approved):
I have re-reviewed the changes and it looks good to me and ready to merge!

And wow, huge speed up with the tetris environment after all the changes! :heart_eyes:

AlexandraVolokhova reviewed (approved):
Looks good to me too!

carriepl commented:
> I have re-reviewed the changes and it looks good to me and ready to merge!
> 
> And wow, huge speed up with the tetris environment after all the changes! 😍

Yeah, the speedups from your mask PR and mine are fully additive so we end up with some experiments on Tetris that get about a 3x speedup! The crystal env is also pretty limited by the speed at which we compute the masks so we might see a 2-3x speedup there also.

With all those PRs merged, I'll re-profile the code to see the changes.
